### PR TITLE
Dev -> master

### PR DESCRIPTION
**Pull request by michaelwangbh**
* Updated README for instructions
* Changed the project name

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+## This example shows how to implement the login, session and form feature using the Microservices architecture design.
```
Comment (chenlica) [inline README.md]: "feature" -> "features". "Microservices" -> "Micro-service"

File: README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+### To run this example:
+1. Clone this repository using `git clone https://github.com/UCI-Chenli-teaching/2019w-project2-login-cart-example.git`
+2. Open Eclipse -> File -> import -> under "Maven" -> "Existing Maven Projects" -> Click "Finish".
+3. For "Root Directory", click "Browse" and select this repository's folder. Click "Finish".
+4. You can run this project on Tomcat now. The default username is `anteater` and password is `123456` .
+
+### Features
+1. This example application allows you to login with the username and password provided above.
+2. When you land on the welcome page, it will show your current session ID and last access time. 
+3. It also simulates a shopping cart feature. When you type in items that you want to store in the current seesion and then click `add`, the web page will show the a list of items that consist of your previous items and the one you just added. When you refresh the page and add more items, the list that the web page shows will contain the all items that you have added to the list in this session.
```
Comment (chenlica) [inline README.md]: "the all" -> "all the"

File: README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+   - In `Filter`, all requests will pass through the `doFilter` function.
+   - `LoginFilter` first checks if the request is `login.html`, `login.js`, or `api/login`, which are the URL patterns we mapped to `LoginServlet.java` that are allowed to access without login.
+   - It then checks if the user has logged in to the current session. If so, it redirects the user to the requested URL and if otherwise,`login.html` .
+  
+  `IndexServlet.java` enables you to see your current session information, last access time and a list of items that you added through a `form` in that session. The `IndexServlet.java` has two methods, `doPost` and `doGet`.
+  * The `doPost` method is invoked when you have HTTP Post requests through the api `/api/index`, which lands on `index.html` through `index.js`.
+    * It first gets the session ID, overrides the last access time, and writes these values in the JSON Object that is sent through `response`. 
+    * Next, `index.js` shows the content in the response by an `ajax` call through `jQuery`, which appears in the `index.html`.
```
Comment (chenlica) [inline README.md]: Remove "the" before "index.html"

File: README.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+  * The `doGet` method is invoked with HTTP Get requests, and it is responsible for the item cart feature.    
+    * First, it gets the session ID and the list of items from current session.
```
Comment (chenlica) [inline README.md]: "from current" -> "from the current"

File: README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    * If there is no such array of items, it will create an empty array and add the item that the user had typed in.
```
Comment (chenlica) [inline README.md]: Remove "had"

File: WebContent/login.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Submit the form content with POST method
+ * @param formSubmitEvent
+ */
+function submitLoginForm(formSubmitEvent) {
+    console.log("submit login form");
+    /**
+     * Important: Disabling the default action of submit
+     * will cause the page to refresh. See jQuery reference
```
Comment (chenlica) [inline WebContent/login.js]: This sentence "Disabling the default action of submit will cause the page to refresh" is a bit unclear.  Can we rephrase it?

File: src/LoginFilter.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    /**
+     * @see Filter#doFilter(ServletRequest, ServletResponse, FilterChain)
+     */
+    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
+        HttpServletRequest httpRequest = (HttpServletRequest) request;
+        HttpServletResponse httpResponse = (HttpServletResponse) response;
+
+        System.out.println("LoginFilter: " + httpRequest.getRequestURI());
+
+        // Check if the URL is allowed to be accessed without logged in
```
Comment (chenlica) [inline src/LoginFilter.java]: "logged" -> "log“

File: src/LoginFilter.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    // You might also want to allow some CSS files, etc..
+    private boolean isUrlAllowedWithoutLogin(String requestURI) {
+        requestURI = requestURI.toLowerCase();
+
+        return requestURI.endsWith("login.html") || requestURI.endsWith("login.js")
+                || requestURI.endsWith("api/login");
+    }
+
+    /**
+     * We need to have these functions because current class implements Filter.
```
Comment (chenlica) [inline src/LoginFilter.java]: Rephrase this comment.

File: README.md (hunk 1 above)
Comment (dayuebai) [reply README.md]: "Microservices" is correct. https://en.wikipedia.org/wiki/Microservices

File: README.md [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+  - In line `47`, it sets up an event listener for the form `submit` action and binds the action to the `submitLoginForm` function. 
+  - The `submitLoginForm` function disables the default form action and sends HTTP POST requests to the backend.
+  - The `handleLoginResult` function parses the JSON data that is sent from the backend. If login is successful, `login.js` redirects to the `index.html` page. If login fails, it shows appropriate error messages.
+
+
+`LoginServlet.java` handles the login requests. It contains the following functionalities:
+  - It gets the username and password from the parameters.
+  - It verifies the username and password.
+  - If login succeeds, it puts the `User` object in the session. Then it sends back a JSON response: `{"status": "success", "message": "success"}` .
+  - If login fails, the JSON response will be: `{"status": "success", "message": "incorrect password"}` or `{"status": "success", "message": "user <username> doesn't exist"}`.
```
Comment (Yicong-Huang) [inline README.md]: status "success" for login fails?

File: WebContent/index.js [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+    success: (resultDataString) => {
+        resultDataJson = JSON.parse(resultDataString);
+        $("#sessionID").text("Session ID: " + resultDataJson["sessionID"]);
+        $("#lastAccessTime").text("Last access time: " + resultDataJson["lastAccessTime"]);
+    }
+});
+
+jQuery("#cart").submit((event) => {
+    event.preventDefault();
+    jQuery.get(
```
Comment (Yicong-Huang) [inline WebContent/index.js]: should we match the style of the `jQuery.get` call with the previous `$.ajex` to enhance consistency? or add comments to show that they are the same but different formet/style to write.

File: WebContent/index.js [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        "api/index",
+        jQuery("#cart").serialize(),
+        (resultDataString) => {
+            var resultArray = resultDataString.split(",");
```
Comment (Yicong-Huang) [inline WebContent/index.js]: avoid using `var`, use `let` or `const` instead

File: WebContent/login.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+function submitLoginForm(formSubmitEvent) {
+    console.log("submit login form");
+    /**
+     * When users click the submit button, the browser will not direct
+     * users to the url defined in HTML form. Instead, it will call this
+     * event handler when the event is triggered.
+     */
+    formSubmitEvent.preventDefault();
+
+    jQuery.post(
```
Comment (Yicong-Huang) [inline WebContent/login.js]: same as the previous `jQuery.get`

File: src/IndexServlet.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Date;
+
+@WebServlet(name = "IndexServlet", urlPatterns = "/api/index")
```
Comment (Yicong-Huang) [inline src/IndexServlet.java]: add comments to explain what is `@WebServlet` doing? Annotation might be a hard topic for java beginners

File: src/LoginServlet.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+
+//
```
Comment (Yicong-Huang) [inline src/LoginServlet.java]: same here, what was the comment?

**Pull request merged**